File: Codice/sgad/servertier/src/main/scala/sgad/servertier/dataaccess/data/userdata/OwnedResource.scala
/**
 * FILE: AuthenticationData.scala
 * PERCORSO /Codice/sgad/servertier/src/main/scala/sgad/servertier/dataaccess/data/userdata
 * DATA CREAZIONE: 19 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-19 - Creazione della classe - Segantin Fabio
 */
package sgad.servertier.dataaccess.data.userdata

import sgad.servertier.dataaccess.data.shareddata.Resource

/**
 * Classe che rappresenta una risorsa posseduta da un utente.
 * @constructor
 * @param resource Il tipo di risorsa.
 * @param quantity La quantità di risorsa posseduta.
 */
class OwnedResource(private val resource: Resource, private var quantity: Int) {
	if (resource == null || quantity < 0)
		throw new IllegalArgumentException

	/**
	 * Metodo getter per l'attributo resource.
	 * @return La risorsa posseduta.
	 */
	def getResource = resource

	/**
	 * Metodo getter per l'attributo quantity.
	 * @return La quantità della risorsa in questione
	 */
	def getQuantity = quantity

	/**
	 * Metodo setter per l'attributo quantity.
	 * @param quantity Nuovo valore da settare per la quantità, deve essere maggiore di zero.
	 */
	def setQuantity(quantity: Int): Unit = {
		if (quantity < 0)
			throw new IllegalArgumentException
		this.quantity = quantity
	}

	/**
	 * Metodo utilizzato per il confronto.
	 * @param other L'oggetto con cui fare il confronto.
	 * @return True o false in base all'esito.
	 */
	override def equals(other: Any): Boolean = {
		if (!other.isInstanceOf[OwnedResource])
			false
		else
			(resource == other.asInstanceOf[OwnedResource].resource
				&& quantity == other.asInstanceOf[OwnedResource].quantity)
	}

	/**
	 * Metodo getter per la key.
	 * @return La chiave della risorsa.
	 */
	def getKey = resource.getKey
}

File: Codice/sgad/clienttier/model/personaldata/OwnedResource.js
/**
 * FILE: OwnedResource.js
 * PERCORSO /Codice/sgad/clienttier/model/personaldata/OwnedResource.js
 * DATA CREAZIONE: 18 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-18 - Creazione della classe - Gatto Francesco
 */

//Eredita da Observable.
OwnedResource.prototype = new Observable();

//Costruttore della classe.
OwnedResource.prototype.constructor = OwnedResource;
/**
 * Classe che rappresenta una risorsa posseduta da un utente.
 * @extends Observable
 * @constructor
 * @param {int} quantity La quantità di risorsa posseduta.
 * @param {Resource} resource Il tipo di risorsa.
 */
function OwnedResource(quantity, resource) {
	/**
	 * La quantità di risorsa posseduta.
	 * @type {int}
	 * @private
	 */
	this.quantity = quantity;
	/**
	 * Il tipo di risorsa.
	 * @type {Resource}
	 * @private
	 */
	this.resource = resource;
}

/**
 * Metodo getter per quantity.
 * @return {int} La quantità della risorsa.
 */
OwnedResource.prototype.getQuantity = function() {
    return this.quantity;
};

/**
 * Metodo setter per quantity.
 * @param {int} quantity La quantità da impostare.
 * @return {void}
 */
OwnedResource.prototype.setQuantity = function(quantity) {
	this.quantity = quantity;
	this.notify();
};

/**
 * Metodo getter per resource.
 * @return {Resource} La risorsa.
 */
OwnedResource.prototype.getResource = function() {
    return this.resource;
};

/**
 * Metodo getter per key.
 * @return {string} La chiave.
 */
OwnedResource.prototype.getKey = function() {
	return this.resource.getKey();
};

/**
 * Ridefinizione del metodo valueOf per OwnedResource.
 * @override
 * @return {string} Identificativo per la risorsa posseduta.
 */
OwnedResource.prototype.valueOf = function() {
    return  JSON.stringify({resource : this.getResource(), quantity : this.getQuantity()});
};

/**
 * Fornisce lo stato della risorsa.
 * @override
 * @return {Number} Lo stato della risorsa.
 */
OwnedResource.prototype.getState = function() {
	return this.quantity;
};